### PR TITLE
1590-We-should-use-a-cleaning-strategy-for-the-generator

### LINKED ISSUE #1590: We should use a cleaning strategy for the generator
The cleaning strategy removes or not parts of the code generated

### PR DESCRIPTION
**Pull request by jecisc**
Add cleaning strategies for MM regeneration.

- No cleaning
- Total cleaning (regenerate all)
- Partial cleaning (we save what was not generated, clean, regenerate and add back what the user previously added to the MM)

Fixes #1590

### CODE REVIEW

File: src/Famix-MetamodelBuilder-Core/FmxBanana.class.st [hunk 1]
```diff
@@ -0,0 +1,8 @@
+"
+I'm a banana
```
Comment (jecisc) [inline src/Famix-MetamodelBuilder-Core/FmxBanana.class.st]: @badetitou Is there a reason you committed this class?

**Comment by jecisc** [comment]
Hum... I tried to regenerate the Java model with that and it fail :(

File: src/Famix-MetamodelBuilder-Core/FmxBanana.class.st (hunk 1 above)
Comment (badetitou) [reply src/Famix-MetamodelBuilder-Core/FmxBanana.class.st]: I wanna say no

File: src/Famix-MetamodelBuilder-Core/FamixMetamodelGeneratorCleanGeneratedStrategy.class.st [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+FamixMetamodelGeneratorCleanGeneratedStrategy >> cleanBeforeGeneration: aMetamodelGenerator [
+	aMetamodelGenerator generatedPackage ifNil: [ ^ self ].
+
+	"We need to select the classes because RPackage return also traits alongside with its classes. We use #defineClasses and not #classes to not get extensions."
+	aMetamodelGenerator generatedPackage definedClasses
+		do: [ :class | 
+			class isMetamodelEntity
+				ifTrue: [
+					addedMethods addAll: (class localMethods , class class localMethods reject: [ :method | method pragmas anySatisfy: [ :pragma | pragma keyword = #generated ] ] thenCollect: [ :m | self methodInfoFor: m ]).
+					(class traitComposition allTraits reject: #isMetamodelEntity) ifNotEmpty: [ :traits | addedTraitsUsers at: class name put: traits ].
```
Comment (jecisc) [inline src/Famix-MetamodelBuilder-Core/FamixMetamodelGeneratorCleanGeneratedStrategy.class.st]: In development of Moose we now have #customTraits

**Comment by jecisc** [comment]
We'll do it in another way.

**Pull request closed without merging**